cabal-install/src/Distribution/Client/CmdTest.hs

```diff
@@ -125,6 +127,10 @@ testAction :: NixStyleFlags () -> [String] -> GlobalFlags -> IO ()
 testAction flags@NixStyleFlags{..} targetStrings globalFlags = do
   baseCtx <- establishProjectBaseContext verbosity cliConfig OtherCommand
 
+  -- Multi package coverage is not supported (see cabal#9493)
+  when (length (localPackages baseCtx) > 1 && fromFlagOrDefault False (packageConfigCoverage (projectConfigLocalPackages $ projectConfig baseCtx))) $
```

Comment:
You're right. I'm not entirely sure how to disable coverage for multi-package components precisely. This is not great, but it is a separate commit that we can drop and revert to the behaviour of failing when `hpc markup` is called because of #9493

Reply:
The solution is to do two things
* Filter out, in cabal-install, the libraries which are not in the package with the testsuite
* In Cabal `test`, if the package being tested does not have any libraries (rather than just if it does not have a main library), we fail as we used to.